### ISSUE TITLE
AllocationRef::~AllocationRef throws from a destructor when the CUDA context is already broken, masking the original error with std::terminate

### ISSUE THREAD

**Issue by yuweih205** [pytorch/pytorch#195824]
## Summary

When a distributed training job dies for an unrelated reason (e.g. `torch.OutOfMemoryError`), `AllocationRef::~AllocationRef()` in `torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu` can throw during teardown. Because this happens **inside a destructor**, it triggers `std::terminate()` and the original, actionable error is replaced by an opaque one:

```
terminate called after throwing an instance of 'c10::Error'
  what():  CUDA driver error: invalid argument
Exception raised from ~AllocationRef at .../symm_mem/CUDASymmetricMemory.cu:71
frame #3: c10d::symmetric_memory::CUDASymmetricMemory::~CUDASymmetricMemory() + 0xc1
```

The real failure (an OOM, a config `ValueError`, …) is still in the logs but is *after* / interleaved with dozens of these aborts across ranks, so it is easy to miss entirely.

## Why this looks like a bug

```cpp
AllocationRef::~AllocationRef() {
  if (is_finalizing()) {
    return;
  }
  c10::cuda::CUDAGuard guard(device_idx);
  C10_CUDA_CHECK(cudaDeviceSynchronize());
#if !defined(USE_ROCM) && defined(PYTORCH_C10_DRIVER_API_SUPPORTED)
  // Leak the cuda allocations during static deinitialization
  auto driver_api = c10::cuda::DriverAPI::get();
  C10_CUDA_DRIVER_CHECK(
      driver_api->cuMemUnmap_(reinterpret_cast<CUdeviceptr>(ptr), block_size));
  ...
  C10_CUDA_DRIVER_CHECK(driver_api->cuMemRelease_(handle));
```

Two observations:

1. The comment says the allocations should be *leaked* during static deinitialization, but the code still calls `cuMemUnmap_` / `cuMemRelease_` and wraps them in `C10_CUDA_DRIVER_CHECK`, which throws on failure.
2. The `is_finalizing()` guard only covers interpreter finalization. In our case the process is being torn down after an exception while the interpreter is *not* yet finalizing, but the CUDA context is already unusable — so the guard does not apply and every CHECK becomes a throw-from-destructor.

Throwing from a destructor is unconditionally fatal in C++, so a cleanup-path failure escalates into `std::terminate` and destroys the diagnostic value of the run.

## Impact

We lost several hours of debugging on a 32-GPU multi-node Megatron-LM training job: every crash surfaced as `CUDA driver error: invalid argument` from this destructor, which sent us chasing context-parallel, NCCL and node-health hypotheses before we found the actual `torch.OutOfMemoryError` earlier in the log. The workaround we now use is to always parse the log *up to* the first `CUDASymmetricMemory` line.

Note that megatron-core creates a global symmetric-memory buffer unconditionally in `parallel_state.initialize_model_parallel()`, so any job that initializes model parallelism has such an object to destroy — even pure-training jobs that never use it.

## Suggested direction

Make the destructor non-throwing, e.g. swallow driver errors and `TORCH_WARN` instead of `C10_CUDA_DRIVER_CHECK`, or extend the "leak instead of free" path to cover the case where the CUDA context is no longer valid (not just interpreter finalization).

## Environment

- torch 2.13.0+cu130 / 2.13.0+cu132, CUDA 13.0, 4× 8×H200, RoCE inter-node
- Megatron-LM core 0.16.0rc0, ms-swift 3.12.0.dev0
- Model: MiMo-V2.5 (MoE, TP4/EP4/PP3), sequence length 26112

## Caveat

I did not manage to reduce this to a standalone repro outside Megatron — the trigger is "process is dying from another error while the CUDA context is already broken", which is awkward to script. Happy to gather more data or test a patch if that helps.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

**Comment by d4l3k** [comment]
@yuweih205 are you planning to send a fix for this? I can help out if not

**Comment by yuweih205** [comment]
Hi @d4l3k, thanks! Yes, I'd like to try — happy to take guidance since it's my first patch here.

Plan: switch the cleanup to the `C10_CUDA_CHECK_WARN` / `C10_CUDA_DRIVER_CHECK_WARN` variants already used elsewhere in this file, plus a try/catch around the remaining throw points (`CUDAGuard` construction, `DriverAPI::get`). Warn and leak — the process is tearing down anyway.

Also worth noting: since filing, the same destructor masked three more unrelated failures for us (a checkpoint `Missing key`, an FA3 backward shape error, a pipeline-config `ValueError`) — so it's not OOM-specific.

I'll send a PR shortly.
